test/test_traversal.py

```diff
@@ -0,0 +1,172 @@
+import uuid
```

Comment:
I know we maintain integration tests make sense when it'd be more complex to set up the mocks for the unit test. But this seems to be way more complex to set up the graph in the db than to just mock it out...

Also totally cool to break out the traversal tests for the main module to reduce clutter but namespace should reflect the location of the code. So: test.psqlgraph.node.test_traversal.py

Reply:
>But this seems to be way more complex to set up the graph in the db than to just mock it out...
I totally agree. I wish the psqlgraph test module of Rowland could be finished soon so we do not have to do anything like this again.

> So: test.psqlgraph.node.test_traversal.py
I think this is a good idea. But I think maybe we should discuss it with the team so that we all follow the same pattern. I will definitely vote for this more explicit pattern.

Reply:
The psqlgraph test module would be no excuse to rely more on integration tests. Integration tests are expensive to run. As we add more and more our entire test suite slows down. Unit tests are cheap and fast; hence, we need to rely on them as much as possible to catch most problems within our code. With a simple refactor your method could easily be turned into a static function: 

`dfs(start_node, edge_predicate, max_depth, edge_pointer)` 

then in the object method just calls dfs(self, ...). We then could test all of it with mock nodes by doing this and then rely on a quick unit test to insure that the logic in dfs is all tested.

Reply:
On a different note, would you be able to highlight in your tests where the following cases are being tested:
ln 247: Max Depth Reached
ln 259: No duplicate node (basically your base case)
ln 261: A revisited node where no update happens
ln 261 (the implied else): A revisited node where update happens.

Reply:
> The psqlgraph test module would be no excuse to rely more on integration tests. Integration tests are expensive to run. As we add more and more our entire test suite slows down. Unit tests are cheap and fast; hence, we need to rely on them as much as possible to catch most problems within our code. With a simple refactor your method could easily be turned into a static function:
> 
> `dfs(start_node, edge_predicate, max_depth, edge_pointer)`
> 
> then in the object method just calls dfs(self, ...). We then could test all of it with mock nodes by doing this and then rely on a quick unit test to insure that the logic in dfs is all tested.

I was following the existing _bfs code. I did not make much change to the tests. I just added a parameter to add my dfs tests with the bsf tests.

Reply:
> On a different note, would you be able to highlight in your tests where the following cases are being tested:
> ln 247: Max Depth Reached
> ln 259: No duplicate node (basically your base case)
> ln 261: A revisited node where no update happens
> ln 261 (the implied else): A revisited node where update happens.

Great questions, I have updated my test to solve those comments. Please take a look! 😄

Reply:
I think my code is thread safe. Going to try future.ThreadPoolExecutor. 
Might save memory and can use more workers.

Reply:
> On a different note, would you be able to highlight in your tests where the following cases are being tested:
> ln 247: Max Depth Reached
> ln 259: No duplicate node (basically your base case)
> ln 261: A revisited node where no update happens
> ln 261 (the implied else): A revisited node where update happens.

To be more specific. 
For 247: Max Depth Reached, please check out test test_traversal_with_max_depth
For 259: No duplicate node (basically your base case). Now I check the length and the set of the returned list.
For 261: This is a little more complicated. Now I have(part of the test graph):

```
        root
        ^    ^
       /       \
foo3  <- test1
    ^           ^
    |            |
test6 -> test4
```
Foo3 should be visited before test1, then:
test1 should be visited twice and an update should happen. Or it will give the wrong level in test_traversal_with_max_depth
test6 should be visited three times and no update should happen  or it will give the wrong level in test_traversal_with_max_depth

If foo3 is visited after test1. The same logic should hold, just reverse the case where updates should and should not happen.